fix: sort_batch function unsupported mixed types with list

JasonLi-cn opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes https://github.com/apache/arrow-datafusion/issues/9418

## Rationale for this change

Support SQL:

```sql
SELECT * FROM table ORDER BY name, book_list
```

where  the`book_list` column is DataType::List .

BTW, I wanted to modify the `lexsort_to_indices` function in [arrow-rs](https://github.com/apache/arrow-rs/compare/master...JasonLi-cn:arrow-rs:fix/lexsort_to_indices_bug
), but after bench, I found that the `Rows` based implementation  was better.

Benchmark on `Apple M1 16G`: 

```shell
lexsort_to_indices([i32_opt, i32_list]): 4096
                        time:   [440.80 µs 442.16 µs 443.63 µs]
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

lexsort_rows([i32_opt, i32_list]): 4096
                        time:   [432.53 µs 434.24 µs 436.16 µs]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

lexsort_to_indices([i32_opt, i32_list]): 32768
                        time:   [4.3801 ms 4.3925 ms 4.4059 ms]
Found 4 outliers among 100 measurements (4.00%)
  4 (4.00%) high mild

lexsort_rows([i32_opt, i32_list]): 32768
                        time:   [4.1159 ms 4.1266 ms 4.1409 ms]
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe

lexsort_to_indices([i32_opt, i32_list_opt]): 4096
                        time:   [433.11 µs 434.57 µs 436.12 µs]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

lexsort_rows([i32_opt, i32_list_opt]): 4096
                        time:   [472.62 µs 473.97 µs 475.40 µs]
Found 7 outliers among 100 measurements (7.00%)
  5 (5.00%) high mild
  2 (2.00%) high severe

lexsort_to_indices([i32_opt, i32_list_opt]): 32768
                        time:   [4.2414 ms 4.2460 ms 4.2508 ms]
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

lexsort_rows([i32_opt, i32_list_opt]): 32768
                        time:   [4.3238 ms 4.3306 ms 4.3376 ms]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

lexsort_to_indices([i32_list_opt, i32_opt]): 4096
                        time:   [533.95 µs 537.64 µs 543.46 µs]
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

lexsort_rows([i32_list_opt, i32_opt]): 4096
                        time:   [418.25 µs 421.96 µs 426.30 µs]
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) high mild
  3 (3.00%) high severe

lexsort_to_indices([i32_list_opt, i32_opt]): 32768
                        time:   [5.2264 ms 5.2651 ms 5.3096 ms]
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) high mild
  3 (3.00%) high severe

lexsort_rows([i32_list_opt, i32_opt]): 32768
                        time:   [3.7982 ms 3.8031 ms 3.8090 ms]
Found 8 outliers among 100 measurements (8.00%)
  3 (3.00%) high mild
  5 (5.00%) high severe

lexsort_to_indices([i32, str_list(4)]): 4096
                        time:   [649.14 µs 650.99 µs 652.99 µs]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe

lexsort_rows([i32, str_list(4)]): 4096
                        time:   [464.10 µs 464.63 µs 465.23 µs]
Found 17 outliers among 100 measurements (17.00%)
  2 (2.00%) high mild
  15 (15.00%) high severe

lexsort_to_indices([i32, str_list(4)]): 32768
                        time:   [6.2407 ms 6.2484 ms 6.2582 ms]
Found 13 outliers among 100 measurements (13.00%)
  7 (7.00%) high mild
  6 (6.00%) high severe

lexsort_rows([i32, str_list(4)]): 32768
                        time:   [4.4957 ms 4.5017 ms 4.5091 ms]
Found 11 outliers among 100 measurements (11.00%)
  7 (7.00%) high mild
  4 (4.00%) high severe

lexsort_to_indices([str_list(4), i32]): 4096
                        time:   [757.10 µs 758.69 µs 760.53 µs]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

lexsort_rows([str_list(4), i32]): 4096
                        time:   [443.60 µs 444.90 µs 446.39 µs]
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe

lexsort_to_indices([str_list(4), i32]): 32768
                        time:   [7.4594 ms 7.4905 ms 7.5280 ms]
Found 11 outliers among 100 measurements (11.00%)
  6 (6.00%) high mild
  5 (5.00%) high severe

lexsort_rows([str_list(4), i32]): 32768
                        time:   [4.0570 ms 4.0609 ms 4.0652 ms]
Found 11 outliers among 100 measurements (11.00%)
  7 (7.00%) high mild
  4 (4.00%) high severe

lexsort_to_indices([i32, str_list_opt(4)]): 4096
                        time:   [586.61 µs 587.43 µs 588.24 µs]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

lexsort_rows([i32, str_list_opt(4)]): 4096
                        time:   [551.36 µs 552.57 µs 553.99 µs]
Found 11 outliers among 100 measurements (11.00%)
  6 (6.00%) high mild
  5 (5.00%) high severe

lexsort_to_indices([i32, str_list_opt(4)]): 32768
                        time:   [5.5504 ms 5.5571 ms 5.5651 ms]
Found 8 outliers among 100 measurements (8.00%)
  4 (4.00%) high mild
  4 (4.00%) high severe

lexsort_rows([i32, str_list_opt(4)]): 32768
                        time:   [5.0307 ms 5.0358 ms 5.0412 ms]
Found 12 outliers among 100 measurements (12.00%)
  3 (3.00%) low mild
  7 (7.00%) high mild
  2 (2.00%) high severe

lexsort_to_indices([str_list_opt(4), i32]): 4096
                        time:   [774.23 µs 775.50 µs 776.79 µs]
Found 9 outliers among 100 measurements (9.00%)
  3 (3.00%) low mild
  3 (3.00%) high mild
  3 (3.00%) high severe

lexsort_rows([str_list_opt(4), i32]): 4096
                        time:   [486.96 µs 488.71 µs 490.63 µs]
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe

lexsort_to_indices([str_list_opt(4), i32]): 32768
                        time:   [7.2347 ms 7.2419 ms 7.2502 ms]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

lexsort_rows([str_list_opt(4), i32]): 32768
                        time:   [4.2789 ms 4.2883 ms 4.2988 ms]
Found 12 outliers among 100 measurements (12.00%)
  3 (3.00%) high mild
  9 (9.00%) high severe

lexsort_to_indices([i32, i32_list, str(16)]): 4096
                        time:   [349.27 µs 351.48 µs 353.60 µs]
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

lexsort_rows([i32, i32_list, str(16)]): 4096
                        time:   [411.21 µs 413.01 µs 414.80 µs]
Found 17 outliers among 100 measurements (17.00%)
  4 (4.00%) low severe
  5 (5.00%) low mild
  5 (5.00%) high mild
  3 (3.00%) high severe

lexsort_to_indices([i32, i32_list, str(16)]): 32768
                        time:   [3.5240 ms 3.5307 ms 3.5373 ms]

lexsort_rows([i32, i32_list, str(16)]): 32768
                        time:   [3.8712 ms 3.8813 ms 3.8926 ms]
Found 10 outliers among 100 measurements (10.00%)
  5 (5.00%) high mild
  5 (5.00%) high severe

lexsort_to_indices([i32_opt, i32_list_opt, str_opt(50)]): 4096
                        time:   [458.24 µs 459.22 µs 460.20 µs]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe

lexsort_rows([i32_opt, i32_list_opt, str_opt(50)]): 4096
                        time:   [446.95 µs 447.50 µs 448.09 µs]
Found 8 outliers among 100 measurements (8.00%)
  4 (4.00%) low mild
  2 (2.00%) high mild
  2 (2.00%) high severe

lexsort_to_indices([i32_opt, i32_list_opt, str_opt(50)]): 32768
                        time:   [4.5581 ms 4.5668 ms 4.5766 ms]
Found 16 outliers among 100 measurements (16.00%)
  5 (5.00%) high mild
  11 (11.00%) high severe

lexsort_rows([i32_opt, i32_list_opt, str_opt(50)]): 32768
                        time:   [4.1926 ms 4.1984 ms 4.2049 ms]
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe
```


## What changes are included in this PR?

upgrade `sort_batch` function.

## Are these changes tested?

Yes.

## Are there any user-facing changes?

No.

alamb commented:
> BTW, I wanted to modify the lexsort_to_indices function in [arrow-rs](https://github.com/apache/arrow-rs/compare/master...JasonLi-cn:arrow-rs:fix/lexsort_to_indices_bug), but after bench, I found that the Rows based implementation was better.

I don't understand this statement -- it says you found the Rows based implementation was better but this PR proposes adding multi-column `lexsort_to_indices` in DataFusion.

Wouldn't it be better to switch DataFusion to use a `Rows` based sort in this case?

alamb reviewed (commented):
Thank you @JasonLi-cn -- this is a good idea

I filed  https://github.com/apache/arrow-datafusion/issues/9418 to track this feature and updated this PR to mention it

I think this PR needs:
1. SQL level end to end tests with multi column sorting with arrays -- perhaps you can adapt the example in https://github.com/apache/arrow-datafusion/issues/9418  and add an sqllogictest: https://github.com/apache/arrow-datafusion/tree/main/datafusion/sqllogictest
2. Since this proposal actually changes all multi-column sorts (not just ones with `List`s) I think we need to run DataFusion benchmarks to make sure it isn't decreasing performance. 

Another approach we could potentially take is to only run this new codepath if there is a list (and there is more than one column)

JasonLi-cn commented:
> Thank you @JasonLi-cn -- this is a good idea
> 
> I filed #9418 to track this feature and updated this PR to mention it
> 
> I think this PR needs:
> 
> 1. SQL level end to end tests with multi column sorting with arrays -- perhaps you can adapt the example in [Support multi column sorting with lists #9418](https://github.com/apache/arrow-datafusion/issues/9418)  and add an sqllogictest: https://github.com/apache/arrow-datafusion/tree/main/datafusion/sqllogictest
> 2. Since this proposal actually changes all multi-column sorts (not just ones with `List`s) I think we need to run DataFusion benchmarks to make sure it isn't decreasing performance.
> 
> Another approach we could potentially take is to only run this new codepath if there is a list (and there is more than one column)

1. I have added some SQL level end to end tests.
2. I refactor `sort_batch`: using RowConverter only when sorting multiple columns that contain a List

alamb reviewed (approved):
Thank you @JasonLi-cn  -- looks great to me. I filed some upstream tickets to track the gap in arrow-rs better.

tustvold commented on datafusion/physical-plan/src/sorts/sort.rs:
```diff
@@ -587,7 +590,13 @@ pub(crate) fn sort_batch(
         .collect::<Result<Vec<_>>>()?;
 
-    let indices = lexsort_to_indices(&sort_columns, fetch)?;
+    let indices = if is_multi_column_with_lists(&sort_columns) {
```
FWIW I would anticipate the row format being faster for almost all multi column sorts

alamb replied:
> FWIW I would anticipate the row format being faster for almost all multi column sorts

My suspicion is that only for large variable length stringss would this approach be faster

tustvold replied:
Certainly when I benchmarked this a while back the cutoff was a lot lower than I was anticipating

alamb commented:
Thanks again @JasonLi-cn
